service/controller/v2/resource/node/evict.go

```diff
@@ -0,0 +1,61 @@
+package node
+
+import (
+	"time"
+
+	"github.com/giantswarm/backoff"
+	"github.com/giantswarm/microerror"
+	"k8s.io/api/core/v1"
+	"k8s.io/api/policy/v1beta1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	apismetav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+)
+
+const (
+	waitInterval = time.Millisecond * 500 // check every 500ms
+	waitTimeout  = time.Minute * 10       // timeout after 2 min
+)
+
+// evict pod from node
+func EvictPod(k8sClient kubernetes.Interface, pod v1.Pod) error {
+	var deleteGracePeriod int64 = 60
+	if pod.DeletionGracePeriodSeconds != nil && *pod.DeletionGracePeriodSeconds > 0 {
+		deleteGracePeriod = *pod.DeletionGracePeriodSeconds
+	}
+	deleteOptions := &apismetav1.DeleteOptions{
+		GracePeriodSeconds: &deleteGracePeriod,
+	}
+	eviction := &v1beta1.Eviction{
+		ObjectMeta: apismetav1.ObjectMeta{
+			Name:      pod.Name,
+			Namespace: pod.Namespace,
+		},
+		DeleteOptions: deleteOptions,
+	}
+	// evict pod
+	err := k8sClient.PolicyV1beta1().Evictions(eviction.Namespace).Evict(eviction)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	getOpts := apismetav1.GetOptions{}
+	// wait for successful termination
+	b := backoff.NewConstant(waitTimeout, waitInterval)
+	o := func() error {
+		p, err := k8sClient.CoreV1().Pods(pod.Namespace).Get(pod.Name, getOpts)
+		if apierrors.IsNotFound(err) || (p != nil && p.ObjectMeta.UID != pod.ObjectMeta.UID) {
+			// pod is no longer in api, we can exit
+			return nil
```

Comment:
I think the approach of blocking the resource here is wrong. We should aim for short running resources. I would suggest to initiate the eviction and cancel the reconciliation without setting the CR status to drained. In the next loop we should check if all pods are evicted and then set the CR status accordingly.

Reply:
WOuld it than make sense to introduce new state into CR? Or where would it be a good place to store information that we created eviction and now we are waiting for pod termination so we can act accordingly in the next loop.

Reply:
I do not know how the eviction API works yet. What happens when we evict multiple times? Does k8s provide information that we already evicted? Then we would not need to track such state by ourselves. We could maybe just evict again and act based on the response.

Reply:
@calvix can't you just check if the pod is evicted in the k8s api?

Reply:
Well, each time this loop run there will be a call for listing all pods on the node, but some pods like daemonSets can be scheduled on nodes that have `unschedulable` status.

If we ignore daemonSets than we don't have to save any status but just repeat the loop until no more pods are on node anymore. That could work.

In fact, `kubectl drain` also ignores daemonSet pods (by default it will exit with an error, but with arg `--ignore-daemonsets` it just ignore them)

Reply:
Are you afraid about daemonsets created by customers?

Reply:
Well, daemonSets are not usually  used for running workloads, but more for the operational thing ( ie calico, monitoring  and other.) So if some customer does some stupid thing and uses daemonSet for their web app then there is nothing we can do.

We can always say that we drain nodes same as `kubectl` drain does so we are synced with the community and it's their fault they run bad podController.